fix: match Juju's action name validation rules (#2723)
canonical/charmcraft: docs/reference/files/actions-yaml-file.rst

.. _actions-yaml-file:

.. _meta::

>     :description: Reference for the ``actions.yaml`` file, including the structure, keys, and naming rules for declaring Juju actions that a charm exposes to operators.

Reviewer: ```suggestion
.. meta::
    :description: Reference for the ``actions.yaml`` file, with descriptions for each of the keys used to declare a charm's supported actions.
```
Tried to align this with the project file description, but I'm not as picky about this one. Let me know what you think.

Author: Seems fine to me, thanks. For the most part people shouldn't need to look at this file anyway, unless they are Juju devs. Although charmcraft.yaml does defer to it a bit.

**Structure:**

*Name:* The name of the key (``<action name>``) is defined by the charm
author. It must match the regular expression
``^[a-z0-9](?:[a-z0-9-]*[a-z0-9])?$`` (lowercase ASCII letters and digits,
with internal hyphens permitted) and must not collide with a Python
:external+python:ref:`keyword <keywords>`. Underscores are not permitted;
hyphens in the action name are mapped to underscores in the Python event

> handler.

Reviewer 2: I think it would be best to leave out the regex ([compare to Snapcraft](https://documentation.ubuntu.com/snapcraft/stable/reference/snapcraft-yaml/#name)) and write this in plain English:

```suggestion
author. It must consist of lowercase ASCII letters, numbers, and hyphens (-). The first
character can't be a hyphen. It must not contain a 
:external+python:ref:`Python keyword <keywords>`. It can't contain underscores (_).
```

Author: Maybe this?

```suggestion
author. It must consist only of lowercase ASCII letters, numbers, and hyphens (-).
A hyphen cannot be the first or last character. It must not be a
:external+python:ref:`Python keyword <keywords`.
```

The underscores bit shouldn't really be needed, since it's already said which characters are allowed, and an underscore was not there. I think it is important that a hyphen can't be the last character, and it's more accurate to say that it can't *be* a keyword, rather than *contain* one, since you certainly can have a "fork" action, even though that contains "for".

Reviewer 2: Your text addresses the nuances much better than mine! I have some followup suggestions ("can't" and "consist entirely"):

```suggestion
author. It must consist only of lowercase ASCII letters, numbers, and hyphens (-).
A hyphen can't be the first or last character. It must not consist entirely of a
:external+python:ref:`Python keyword <keywords`.
```
